# 📊 Data Understanding

In this stage, the objective is to gain a comprehensive understanding of the dataset before performing any data cleaning or exploratory analysis.

This process includes reviewing the dataset structure, identifying data types, assessing data quality, and understanding the business meaning of each variable.

---

## Dataset Overview

Before conducting the analysis, it is important to understand the dataset structure, available variables, and overall data quality.

The dataset contains hotel booking records for both City Hotels and Resort Hotels.

---

## Dataset Information

Source :
Kaggle - Hotel Booking Demand

Number of Rows :
119,390

Number of Columns :
32

Observation Unit :
Each row represents one hotel booking.

Target Variable :
is_canceled

---

## Data Preview

The first step is to inspect the dataset by displaying several rows. This helps us understand the available variables and verify that the dataset has been loaded correctly.

In [1]:
import pandas as pd

df = pd.read_csv('../data/hotel_bookings.csv')

df.head()

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


## Dataset Shape

Understanding the size of the dataset is important before performing any analysis. This section provides the number of observations (rows) and variables (columns) available.

In [2]:
rows, cols = df.shape

print(f"Number of Rows    : {rows:,}")
print(f"Number of Columns : {cols}")

Number of Rows    : 119,390
Number of Columns : 32


The dataset contains 119,390 booking records and 32 variables. The number of observations is sufficiently large to perform reliable exploratory analysis and identify meaningful business patterns.

## Data Types

Before analysis, it is necessary to understand the data types of each variable to determine the appropriate preprocessing and analytical methods.

In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  str    
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  str    
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal                       

## Statistical Summary

Descriptive statistics provide an overview of the distribution of numerical variables. This helps identify unusual values, understand customer behavior, and determine whether further investigation is needed before analysis.

In [4]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
is_canceled,119390.0,0.370416,0.482918,0.00,0.00,0.000,1.0,1.0
lead_time,119390.0,104.011416,106.863097,0.00,18.00,69.000,160.0,737.0
arrival_date_year,119390.0,2016.156554,0.707476,2015.00,2016.00,2016.000,2017.0,2017.0
arrival_date_week_number,119390.0,27.165173,13.605138,1.00,16.00,28.000,38.0,53.0
arrival_date_day_of_month,119390.0,15.798241,8.780829,1.00,8.00,16.000,23.0,31.0
stays_in_weekend_nights,119390.0,0.927599,0.998613,0.00,0.00,1.000,2.0,19.0
stays_in_week_nights,119390.0,2.500302,1.908286,0.00,1.00,2.000,3.0,50.0
adults,119390.0,1.856403,0.579261,0.00,2.00,2.000,2.0,55.0
children,119386.0,0.103890,0.398561,0.00,0.00,0.000,0.0,10.0
babies,119390.0,0.007949,0.097436,0.00,0.00,0.000,0.0,10.0


### Booking Behavior

Customers typically book their stay around **104 days before arrival**. However, the maximum lead time reaches **737 days**, indicating that some reservations are made nearly two years in advance. Such a wide range suggests that booking behavior varies considerably and that lead time may be an important factor influencing booking cancellations.

Most bookings involve approximately **two adults**, which aligns with typical hotel booking behavior. However, the maximum value of **55 adults** appears unusually high and should be examined further to determine whether it represents a group reservation or an anomaly.

On average, customers make **fewer than one special request** per booking, indicating that additional service requests are relatively uncommon among hotel guests.

### Potential Data Quality Issues

The ADR ranges from **-6.38 to 5400**, indicating the presence of extreme values. Since a negative room rate is generally not realistic, this variable requires further investigation during the data cleaning stage to determine whether these values represent data entry errors or valid business cases.

The `children` variable contains only **4 missing values**. Although the proportion is negligible, these observations should still be reviewed during data cleaning to ensure data consistency.

Only a small portion of bookings is associated with a company, explaining the large number of missing values in the `company` variable. Similarly, although most bookings are linked to travel agents, some records contain missing values in the `agent` variable. These variables require further assessment before selecting an appropriate handling strategy.

### Exceptional Cases

Most bookings are **not placed on a waiting list**, as indicated by the median value of zero. However, a small number of bookings experienced waiting periods of up to **391 days**, suggesting the existence of exceptional booking cases that may warrant further investigation.

### Overall Observation

Based on the descriptive statistics, several key observations can be identified:

- Booking behavior varies considerably, particularly in terms of lead time, suggesting that customers plan their trips differently.
- Several variables contain extreme or potentially unrealistic values (e.g., ADR and number of adults), which should be validated during the data cleaning stage.
- Most bookings involve two adults, with relatively few children, babies, or special service requests.
- Missing values are concentrated in the `company` and `agent` variables, indicating that not all bookings are made through companies or travel agents.
- Overall, the dataset appears sufficiently complete and suitable for further exploratory analysis after addressing data quality issues.